scripts/build.py

```diff
@@ -549,6 +525,38 @@ def save_hashes_as_json(target_filepath, file_hashes):
     file_hashes[relative_filepath] = file_hash
 
 
+def minify_func(source_path, target_path, file_hashes, filename):
+    if filename.endswith('.html'):
+        process_html(source_path, target_path, file_hashes)
+    elif filename.endswith('.css'):
+        _minify(source_path, target_path)
+    elif filename.endswith('.js'):
+        _minify(source_path, target_path)
+    else:
+        shutil.copyfile(source_path, target_path)
+
+
+def _execute_tasks(tasks, batch_size=24):
+    """Starts all tasks and checks the results.
+
+    Runs no more than 'batch_size' tasks at a time.
+    """
+    remaining_tasks = [] + tasks
+    currently_running_tasks = set([])
+
+    while remaining_tasks or currently_running_tasks:
+        if currently_running_tasks:
+            for task in list(currently_running_tasks):
```

Comment:
Instead of having `currently_running_tasks` as a set, use a list instead. Loop over the list using an index and then use`del currently_running_tasks[i]` to remove the task from the list. This prevents converting `currently_running_tasks` to a list many times.

Reply:
Done.

Reply:
Actually, I'm not sure this is the best way to go about it. It was bothering me a bit that we were deleting elements from the list we were iterating over. Though it didn't seem to cause any problems to have `currently_running_tasks` as a list and using del, I didn't notice any performance gains using just wall time -- it was actually around 10 seconds slower on 3 tries. I can believe there will be a situation where the del occurs on the last element of `currently_running_tasks` and there are no more tasks in `remaining_tasks`. Then we will have to wait for another iteration to reset the enumerate index. More frequently, if two tasks in a row are both completed, only the first will be deleted, then the index will shit the second to where the first was and iterate past it. This means there will be another iteration before the second gets cleaned, delaying the spawn of another task. This is likely the cause of the 10 second slowdown. It's possible to do this with a reversed list, but the cost is the same as changing it to a list from a set, so I think it's better as is.

Reply:
You do raise a good point. However, I don't really think transforming sets to lists over and over again is the right approach due to the need to constantly rehash repeatedly. Another idea that I have is to simply replace the pointer to `remaining_tasks.pop()` at the index where the task is completed. The cost is lower and you can let the garbage collector deal with the orphaned task. This would also allow you to remove the latter loop, which will allow you get some more performance gains. This might be helpful [PythonSpeed](https://wiki.python.org/moin/PythonSpeed#Techniques_for_Improving_Performance_and_Scalability) along with the Performance tips links that they have at the bottom. Essentially, you'd want to offload as much as possible to the parts that Python has implemented in C++ so it's way faster.

Reply:
I think it's fine to just change currenty_running_tasks to a list. I'm still uncomfortable with the idea of modifying a list while iterating using for. It's possible to rewrite it with a while, but I'd have to just add more branching logic to determine if I should be expecting to add more remaining tasks. In order to try and get more speedup I ran the script through cProfile and less than .04 seconds is spent on operations on list objects. Honestly, I don't want to have to go through the trouble or changing this, then testing all the changes and making sure it still works for a speedup of something like .02 seconds. This logic is the same as used for the backend test multithreading and seems to be well tested.

Reply:
Okay, we can leave it as is then since we already have an example of this elsewhere. Though this does mean that you should have sleep like the backend tests do. This is probably worth a try since it's not a huge code change (since the api is very similar), but what happens if you use processes instead of threads?

Reply:
Using processes achieved the same result, which makes me think my suspicion that this is IO bound is probably right. The only part of this that's still hard to measure is the subprocess call, as I cannot tell if that is spending time waiting for IO or computing. It should release the GIL so I don't think we get anything for free using processes instead.

Reply:
Interesting find, I was hoping it would be more beneficial since minification does utilize more varied services than the backend tests. In that case, can you push your latest changes so we can close this out?

Reply:
sounds good, thanks!